### ISSUE TITLE
[REVIEW]: pendsim: Developing, Simulating, and Visualizing Feedback Controlled Inverted Pendulum Dynamics

### ISSUE THREAD

**Issue by whedon** [openjournals/jose-reviews#168]
**Submitting author:** @rland93 (<a href="http://orcid.org/0000-0001-5394-2737">Mike Sutherland</a>)
**Repository:** <a href="https://github.com/EASEL-UCI/pendsim" target ="_blank">https://github.com/EASEL-UCI/pendsim</a>
**Version:** v1.2
**Editor:** @moorepants
**Reviewer:** @cooperrc, @repagh
**Archive:** <a href="https://doi.org/10.5281/zenodo.7603937" target="_blank">10.5281/zenodo.7603937</a>

**:warning: JOSE reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSE is currently operating in a "reduced service mode". 

## Status

[![status](https://jose.theoj.org/papers/f2bb0a8f4e8b9b59c955c1cc482a2b5a/status.svg)](https://jose.theoj.org/papers/f2bb0a8f4e8b9b59c955c1cc482a2b5a)

Status badge code:

```
HTML: <a href="https://jose.theoj.org/papers/f2bb0a8f4e8b9b59c955c1cc482a2b5a"><img src="https://jose.theoj.org/papers/f2bb0a8f4e8b9b59c955c1cc482a2b5a/status.svg"></a>
Markdown: [![status](https://jose.theoj.org/papers/f2bb0a8f4e8b9b59c955c1cc482a2b5a/status.svg)](https://jose.theoj.org/papers/f2bb0a8f4e8b9b59c955c1cc482a2b5a)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/rland93/pendsim" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@cooperrc & @repagh, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/jose-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @moorepants know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSE conflict of interest policy](https://github.com/openjournals/joss/blob/jose/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSE code of conduct](https://jose.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/rland93/pendsim">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** v1.2
- [x] **Authorship:** Has the submitting author (@rland93) made substantial contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation? (and documentation is sufficient?)
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [x] **Installation instructions:** Is there a clearly stated list of dependencies? (Ideally these should be handled with an automated package management solution.)
- [x] **Example usage:** Do the authors include examples of how to use the software?
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSE conflict of interest policy](https://github.com/openjournals/joss/blob/jose/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSE code of conduct](https://jose.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/rland93/pendsim">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** v1.2
- [x] **Authorship:** Has the submitting author (@rland93) made substantial contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation? (and documentation is sufficient?)
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [x] **Installation instructions:** Is there a clearly stated list of dependencies? (Ideally these should be handled with an automated package management solution.)
- [x] **Example usage:** Do the authors include examples of how to use the software?
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @cooperrc, @repagh it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/jose-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/jose-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1049

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.12 s (228.9 files/s, 82072.4 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                           9            264            586            919
Jupyter Notebook                 6              0           6920            386
Markdown                         2             65              0            124
reStructuredText                 4             82             98            101
TeX                              1              2              0             29
DOS Batch                        1              8              1             26
make                             1              5              7             16
YAML                             1              6             10             14
CSS                              1              1              0              8
TOML                             1              0              0              3
-------------------------------------------------------------------------------
SUM:                            27            433           7622           1626
-------------------------------------------------------------------------------


Statistical information for the repository '1da1f6afba6c0cc2ccd739dd' was
gathered on 2022/04/01.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Mike Sutherland                 19          2585           2036           29.12
ms                              23          5399           4472           62.21
rland93                          6           954            422            8.67

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Mike Sutherland            1690           65.4          2.5                8.46
rland93                      79            8.3          3.6               41.77
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1515/9781400828739 is OK
- 10.1109/MCSE.2007.55 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/jose-papers/jose.00168/jose.00168/10.21105.jose.00168.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/jose-papers/blob/jose.00168/jose.00168/10.21105.jose.00168.pdf) :page_facing_up: :point_left:

**Comment by moorepants** [comment]
@cooperrc, @repagh, and @rland93 welcome to the review issue. For the reviewers please read through the above information. I was wrong in the other thread on the timeline, it seems that **6 weeks** is the recommended due date. Note that JOSE reviews are a bit different than most academic journal reviews in that they are open here on Github and there can be a conversation among the reviewers, editors, and authors to arrive a final version that meets the standards for publication. You all can start the review when you'd like and I'll jump in to answer questions or help things move along. Thanks again for volunteering your time and expertise!

**Comment by moorepants** [comment]
@whedon remind @reviewer in 4 weeks

**Comment by whedon** [comment]
Reminder set for @reviewer in 4 weeks

**Comment by moorepants** [comment]
Also I believe this submission is the second type we accept "open-source software, created as educational technology or infrastructure" (see https://openjournals.readthedocs.io/en/jose/submitting.html). @rland93, is that correct?

**Comment by whedon** [comment]
:wave: @cooperrc, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @repagh, please update us on how your review is going (this is an automated reminder).

**Comment by moorepants** [comment]
@cooperrc and @repagh can we get an update on the status of your reviews? I see that @repagh has worked through his checklist. Do we expect completion of the first reviews within the 6 weeks deadline? Thanks -- Jason

**Comment by cooperrc** [comment]
> @cooperrc and @repagh can we get an update on the status of your reviews? I see that @repagh has worked through his checklist. Do we expect completion of the first reviews within the 6 weeks deadline? Thanks -- Jason

Thanks for checking in @moorepants. I'm getting back to this today. I'll get my initial thoughts down this week (4/25-4/29). Will have my review done within the 6-week timeframe (before May 13)

**Comment by cooperrc** [comment]
I added an issue (closed) and two PRs to the `pendsim` repo. 

My review is as follows:

The package and the paper are excellent. The one spot that seems to be lacking is a clear path to extend the package (maybe another section for developers). In the paper, @rland93 says the user can implement any control policy with a new class. It would be great to see _how_ to create the new class and _how/if_ the new class could be added to the current repository of controller algorithms. 

Another minor note, the `## Example Notebooks` could be added to the documentation [Tutorials](https://rland93.github.io/pendsim/tutorial.html) for more in-depth examples. 

- [x] https://github.com/rland93/pendsim/issues/11

- [ ] https://github.com/rland93/pendsim/pull/13

- [ ] https://github.com/rland93/pendsim/pull/14

**Comment by rland93** [comment]
Will work on implementing these changes this week. Details:

https://github.com/rland93/pendsim/issues/15
https://github.com/rland93/pendsim/issues/16
https://github.com/rland93/pendsim/issues/17

Thanks!

**Comment by moorepants** [comment]
@repagh and @cooperrc thank you both for the reviews.

@rland93 I see that there is one checkbox missing on both lists above "community guidelines" and both reviewers have left their comments here or on issues on your repository. Please address the issues and when ready you can provide a response here to each reviewer. Let me know if you have any questions. Feel free to interact with the reviewers here also.

**Comment by moorepants** [comment]
@rland93 Can you provide an update on your status in addressing the feedback?

**Comment by moorepants** [comment]
@rland93 Just checking in again here. Can you provide an update on your status in addressing the feedback?

**Comment by labarba** [comment]
@moorepants — I received an email from David A. Copp saying Mike Sutherland (@rland93) has been incommunicado for several months, and if he can be swapped as corresponding author to finish up this submission.

**Comment by moorepants** [comment]
David Copp is currently listed as a co-author on the paper. Although, I do not see any contributions from him in the pendsim repository: https://github.com/rland93/pendsim/graphs/contributors. We would need his github handle to swap him out, of course.

**Comment by EASEL-UCI** [comment]
Hello, this is David Copp. I appreciate the feedback and will happily finish up this submission.

**Comment by moorepants** [comment]
Hi David. The review is present here and as issues on your repository. Any outstanding items on the checklists above and all PRs/issues on your repository need to be addressed as well as providing a response to the reviewers' comments.

**Comment by moorepants** [comment]
Hi @EASEL-UCI, do you plan to complete this paper? We have not heard from you for about a month.

**Comment by EASEL-UCI** [comment]
Yes, definitely. I apologize for the delay. I will need to fork the original repository and make changes there. Will it be an issue to change this submission's repository to the new forked repository?

**Comment by moorepants** [comment]
> Will it be an issue to change this submission's repository to the new forked repository?

I suspect we will be able to fix that. @labarba can you comment?

**Comment by labarba** [comment]
I don't remember dealing with a change in target repo. @openjournals/dev – can you confirm whether we can edit the submission repository?

**Comment by arfon** [comment]
You can yes but I'll need to edit it in the database too. What's the new repo?

**Comment by EASEL-UCI** [comment]
> TY, @arfon. @EASEL-UCI : please report here the URL of the fork you will use for completing this review.

Excellent. Thank you. Please use https://github.com/EASEL-UCI/pendsim

**Comment by arfon** [comment]
Updated.

**Comment by EASEL-UCI** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/jose-papers/jose.00168/jose.00168/10.21105.jose.00168.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/jose-papers/blob/jose.00168/jose.00168/10.21105.jose.00168.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/jose-papers/jose.00168/jose.00168/10.21105.jose.00168.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/jose-papers/blob/jose.00168/jose.00168/10.21105.jose.00168.pdf) :page_facing_up: :point_left:

**Comment by EASEL-UCI** [comment]
The following changes were made:
- Added a section on "Contributing and reporting issues" to README.md
- Added References section to paper.md
- Added discussion of how the package and notebooks can be used in teaching
- Added an example course assignment notebook on designing a PID controller

@cooperrc: Thank you for your feedback and the excellent proposed changes. I believe we have addressed your comments, but if not or if you have other comments or concerns, please let me know. 

@repagh: Thank you for your helpful feedback and for catching errors. Please find our responses to your comments below.

This is the only Python package I am aware of that allows for design, analysis, and visualization of control and estimation algorithms for an inverted pendulum system. Moreover, the example notebooks cover multiple algorithms, whereas other projects may focus solely on PID control, for example. We also added a tutorial on how to create a custom control algorithm (https://rland93.github.io/pendsim/customctl.html).  

Regarding how to use this in teaching, you make an excellent point. We added the following paragraph in the Example Usage section to briefly describe how this package can be used in education, and we added an example assignment notebook. The emphasis here is on augmenting in-class activities or in-lab hardware experiments rather than serving as the sole instructional material on a topic. For example, there are many excellent online resources for learning PID control; we do not try to replicate those. Instead we aim to provide a sandbox that students can easily use to gain experience tuning a PID controller, visualizing the results, and reflecting on the process.

"This design-test-visualize sequence also allows instructors to introduce 
students to new topics and create interactive assignments that can augment 
theoretical class discussions and hardware experiments. Thus, `pendsim` 
may be naturally used in engineering courses related to dynamics and control 
that are taught in any format, including in-person, hybrid, and online settings. 
An example [PID Controller Design Assignment notebook](https://colab.research.google.com/github/EASEL-UCI/pendsim/blob/main/notebooks/PID_controller_design_assignment.ipynb) is 
available to show how students may use `pendsim` in a course assignment."

**Comment by moorepants** [comment]
@EASEL-UCI, thanks for the update.

@cooperrc and @repagh, can you now review the updates and see if it satisfies your concerns? If so or if not, you can comment here with approval for publishing or requests for more adjustments. Please finalize the checklists above if those are now complete.

**Comment by cooperrc** [comment]
My comments and suggestions were addressed and satisfies any concerns I had. 

Thank you for your contribution @EASEL-UCI !

**Comment by moorepants** [comment]
@EASEL-UCI I see some math rendering errors here:
![image](https://user-images.githubusercontent.com/276007/194847314-63ffcdf8-7397-4b05-bcdf-e2c42bb45af2.png)
https://rland93.github.io/pendsim/customctl.html#Model-Predictive-Control-(MPC)

Maybe a pass through to check for these things is worthwhile.

**Comment by moorepants** [comment]
@whedon set v1.1.0 as version

**Comment by whedon** [comment]
OK. v1.1.0 is the version.

**Comment by moorepants** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1515/9781400828739 is OK
- 10.1109/MCSE.2007.55 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.07 s (404.0 files/s, 162147.6 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                           9            270            580            939
Jupyter Notebook                 8              0           7984            564
Markdown                         2             62              0            184
YAML                             1              9             10             44
reStructuredText                 2             25             22             37
TeX                              1              2              0             29
DOS Batch                        1              8              1             26
make                             1              5              7             16
CSS                              1              1              0              8
TOML                             1              0              0              3
-------------------------------------------------------------------------------
SUM:                            27            382           8604           1850
-------------------------------------------------------------------------------


Statistical information for the repository '4b2bd083c821b41ef8f29cb7' was
gathered on 2022/10/10.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Mike Sutherland                 19          2585           2036           29.00
ms                              23          5399           4472           61.95
rland93                          9           997            445            9.05

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Mike Sutherland            1681           65.0          9.3                8.51
rland93                     108           10.8          6.3               25.00
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/jose-papers/jose.00168/jose.00168/10.21105.jose.00168.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/jose-papers/blob/jose.00168/jose.00168/10.21105.jose.00168.pdf) :page_facing_up: :point_left:

**Comment by moorepants** [comment]
There isn't an issues tab on the new repository so I'll leave some things here. The old repository is still referenced in many places, for example:
![image](https://user-images.githubusercontent.com/276007/194848254-308fc7cf-0f63-44b2-b859-6a86c0438186.png)
Do you plan to update these?

**Comment by moorepants** [comment]
You are using a theme that has a dark mode. Many of the figures are unreadable, e.g.:
![image](https://user-images.githubusercontent.com/276007/194848584-f765223e-2426-4ce7-9637-5af18911769f.png)
Maybe disable the dark mode as a quick fix?

**Comment by moorepants** [comment]
Tutorial figure on line 7 has misplaced axes labels:
![image](https://user-images.githubusercontent.com/276007/194848793-99ed3ab0-406b-47fd-8150-d2bc57471720.png)

**Comment by moorepants** [comment]
The tutorial video on the last line does not fit on narrow screens:
![image](https://user-images.githubusercontent.com/276007/194848903-e51ad4cb-fc9c-460f-8446-a04292455590.png)

**Comment by moorepants** [comment]
The `sys.append('../')` lines should not be needed in your examples:
```python
import sys
sys.path.append('../')
%matplotlib notebook
from pendsim import sim, controller, viz
import numpy as np
import matplotlib.pyplot as plt
```
If users install your package, they shouldn't have to do that.

**Comment by moorepants** [comment]
On the animation page there is some unusual directory changing:
```
import os, pathlib
os.chdir(pathlib.Path(globals()['_dh'][0]).parent)
```
This does not seem like good practice.

**Comment by moorepants** [comment]
On the linearization page there are some formatting errors, e.g.:
![image](https://user-images.githubusercontent.com/276007/194849494-5c0ad716-1489-4e9b-8900-efd283090cc4.png)

**Comment by moorepants** [comment]
All of the cells are executed on the PID page:

https://rland93.github.io/pendsim/PID.html

**Comment by moorepants** [comment]
@EASEL-UCI Can you address the various formatting and presentation issues I've mentioned here? This would bring the contents to publishable quality with JOSE. Once that is done, you can cut a new release.

@repagh and @cooperrc, I assume that the "community guidelines" checkbox was not checked by mistake. Can you check that box if you are satisfied with the statement in the README?

**Comment by labarba** [comment]
hi folks! 👋  Happy New Year... 
It looks like this is close to the finish line. We just need some final efforts on your part, @EASEL-UCI – have a look 👀

**Comment by EASEL-UCI** [comment]
@labarba Happy New Year. Thanks very much for the note.
@moorepants Thank you for noticing these remaining issues and for the detailed comments. We've made the following changes and believe we've addressed each of the comments:
- updated documentation so that we no longer reference the old repository
- fixed misplaced axis labels and improved figure labels throughout docs
- removed all sys, os, and path commands from examples
- fixed formatting of bulleted lists on linearization docs page
- executed all cells on PID page
- video scaling improved, and all videos can be expanded to fit a particular screen
- users can disable dark theme to better view figures in docs

The new release v1.2 is up-to-date with these changes.

**Comment by moorepants** [comment]
@whedon set v1.2 as version

**Comment by whedon** [comment]
OK. v1.2 is the version.

**Comment by moorepants** [comment]
@whedon generate pdf

**Comment by moorepants** [comment]
@whedon check references

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/jose-papers/jose.00168/jose.00168/10.21105.jose.00168.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/jose-papers/blob/jose.00168/jose.00168/10.21105.jose.00168.pdf) :page_facing_up: :point_left:

**Comment by moorepants** [comment]
@whedon check repository

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1111

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.12 s (224.5 files/s, 323671.2 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                           9            270            580            939
Jupyter Notebook                 8              0          36076            572
Markdown                         2             62              0            184
YAML                             1              9             10             44
reStructuredText                 2             20             20             32
TeX                              1              2              0             29
DOS Batch                        1              8              1             26
make                             1              5              7             16
CSS                              1              1              0              8
TOML                             1              0              0              3
-------------------------------------------------------------------------------
SUM:                            27            377          36694           1853
-------------------------------------------------------------------------------


Statistical information for the repository '8ecd942a11012574e73252c5' was
gathered on 2023/02/01.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
EASEL Research Group             3             3              3            0.04
Mike Sutherland                 19          2585           2036           28.99
ms                              23          5399           4472           61.93
rland93                          9           997            445            9.05

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
EASEL Research Group          3          100.0          1.1                0.00
Mike Sutherland            1679           65.0         17.5                8.52
rland93                     107           10.7         14.6               25.23
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1515/9781400828739 is OK
- 10.1109/MCSE.2007.55 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by moorepants** [comment]
@EASEL-UCI this still shows documentation for 1.1: https://rland93.github.io/pendsim/index.html Should I be looking elsewhere?

**Comment by EASEL-UCI** [comment]
@moorepants Yes, please see https://easel-uci.github.io/pendsim/index.html

**Comment by moorepants** [comment]
I've checked over everything again and I approve for publishing. @EASEL-UCI can you make the Zenodo archive of your version 1.2.0 and reply here with the DOI to that archive?

**Comment by EASEL-UCI** [comment]
@moorepants Thank you very much! Here is the Zenodo DOI: 10.5281/zenodo.7603937

**Comment by moorepants** [comment]
@whedon set 10.5281/zenodo.7603937 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.7603937" target="_blank">10.5281/zenodo.7603937</a> is the archive.

**Comment by moorepants** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1515/9781400828739 is OK
- 10.1109/MCSE.2007.55 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/jose-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/jose-papers/pull/113

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/jose-papers/pull/113, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by moorepants** [comment]
Hi @labarba, this looks ready from my perspective.

@repagh and @cooperrc, thank you very much for volunteering your time to review this submission.

@EASEL-UCI and @rland93, congrats on having this published!

**Comment by EASEL-UCI** [comment]
@moorepants, @repagh, and @cooperrc: Thank you very for your valuable feedback that greatly improved the software. We appreciate your time!

**Comment by moorepants** [comment]
@labarba, This paper is ready to be published.

**Comment by labarba** [comment]
I notice that both reviewers left the "Community guidelines" item of the checklist unchecked. But they both approved for publication in a later comment, @cooperrc [here](https://github.com/openjournals/jose-reviews/issues/168#issuecomment-1259781500) and @repagh [here](https://github.com/openjournals/jose-reviews/issues/168#issuecomment-1270106675). I will proceed with publication, and mark those items as checked.

**Comment by labarba** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSE!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/jose-papers/pull/115
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/jose.00168](https://doi.org/10.21105/jose.00168)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by labarba** [comment]
Congratulations, @rland93 and @EASEL-UCI, your JOSE paper is published! 🚀 

Big thanks to our Editor: @moorepants, and to the Reviewers: @cooperrc, @repagh — your contributions make this adventure in publishing possible 🙏

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://jose.theoj.org/papers/10.21105/jose.00168/status.svg)](https://doi.org/10.21105/jose.00168)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/jose.00168">
  <img src="https://jose.theoj.org/papers/10.21105/jose.00168/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://jose.theoj.org/papers/10.21105/jose.00168/status.svg
   :target: https://doi.org/10.21105/jose.00168
```
This is how it will look in your documentation:

[![DOI](https://jose.theoj.org/papers/10.21105/jose.00168/status.svg)](https://doi.org/10.21105/jose.00168)

**We need your help!**

Journal of Open Source Education is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future.
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-jose

**Comment by labarba** [comment]
:loudspeaker:  New Reviewer Sign-up Form!
https://forms.gle/Rizd3TcHnQKhrbYY7
